pkg/resources/stream_common.go

```diff
@@ -58,3 +58,79 @@ func DeleteStreamContext(ctx context.Context, d *schema.ResourceData, meta any)
 	d.SetId("")
 	return nil
 }
+
+var atSchema = &schema.Schema{
+	Type:        schema.TypeList,
+	Optional:    true,
+	MaxItems:    1,
+	Description: externalChangesNotDetectedFieldDescription("This field specifies that the request is inclusive of any changes made by a statement or transaction with a timestamp equal to the specified parameter. Due to Snowflake limitations, the provider does not detect external changes on this field."),
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"timestamp": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Specifies an exact date and time to use for Time Travel. The value must be explicitly cast to a TIMESTAMP, TIMESTAMP_LTZ, TIMESTAMP_NTZ, or TIMESTAMP_TZ data type.",
+				ExactlyOneOf: []string{"at.0.timestamp", "at.0.offset", "at.0.statement", "at.0.stream"},
+			},
+			"offset": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Specifies the difference in seconds from the current time to use for Time Travel, in the form -N where N can be an integer or arithmetic expression (e.g. -120 is 120 seconds, -30*60 is 1800 seconds or 30 minutes).",
+				ExactlyOneOf: []string{"at.0.timestamp", "at.0.offset", "at.0.statement", "at.0.stream"},
+			},
+			"statement": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Specifies the query ID of a statement to use as the reference point for Time Travel. This parameter supports any statement of one of the following types: DML (e.g. INSERT, UPDATE, DELETE), TCL (BEGIN, COMMIT transaction), SELECT.",
+				ExactlyOneOf: []string{"at.0.timestamp", "at.0.offset", "at.0.statement", "at.0.stream"},
+			},
+			"stream": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				Description:      "Specifies the identifier (i.e. name) for an existing stream on the queried table or view. The current offset in the stream is used as the AT point in time for returning change data for the source object.",
+				ExactlyOneOf:     []string{"at.0.timestamp", "at.0.offset", "at.0.statement", "at.0.stream"},
+				DiffSuppressFunc: suppressIdentifierQuoting,
+				ValidateDiagFunc: IsValidIdentifier[sdk.SchemaObjectIdentifier](),
+			},
+		},
+	},
+	ConflictsWith: []string{"before"},
+}
+
+var beforeSchema = &schema.Schema{
+	Type:        schema.TypeList,
+	Optional:    true,
+	MaxItems:    1,
+	Description: externalChangesNotDetectedFieldDescription("This field specifies that the request refers to a point immediately preceding the specified parameter. This point in time is just before the statement, identified by its query ID, is completed.  Due to Snowflake limitations, the provider does not detect external changes on this field."),
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
```

Comment:
Tbh I'm not sure if there are any differences, but you could go full recycling mode and define subfields as one variable and use it in both schemas, but this kind of duplication is not that bad.

Reply:
Note that these fields have different ExactlyOneOf, so this needs to be parameterized. We'd only save a couple of lines, which is not much.

Reply:
Yeah I didn't notice 👍